### PR TITLE
Fix GetAffectedFiles failed when encountering a null old commit

### PR DESCRIPTION
**Pull request by nanguanlin6**
When using protected/unprotected file patterns, `git push --set-upstream xxx xxx` will cause an internal server error.
6D9IB}`MLNADEJ%QV`QD(0V
This bug is because when `git push --set-upstream`, a null commit is passed to function GetAffectedFiles as oldCommitID, but git diff cannot compare null commits. So when `oldCommit` is null, we should modify the `oldCommitID` to empty tree commit("4b825dc642cb6eb9a060e54bf8d69288fbee4904") to use `git diff`.

### CODE REVIEW

File: modules/git/diff.go [hunk 1]
```diff
@@ -272,6 +272,10 @@ func CutDiffAroundLine(originalDiff io.Reader, line int64, old bool, numbersOfLi
 
 // GetAffectedFiles returns the affected files between two commits
 func GetAffectedFiles(repo *Repository, oldCommitID, newCommitID string, env []string) ([]string, error) {
+	// If the old commit is the null commit, then we need to use the empty tree
+	if oldCommitID == "0000000000000000000000000000000000000000" {
+		oldCommitID = "4b825dc642cb6eb9a060e54bf8d69288fbee4904"
```
Comment (CaiCandong) [inline modules/git/diff.go]:
```suggestion
	if oldCommitID == EmptySHA {
		oldCommitID = EmptyTreeSHA
	}
```

**Review by delvh — APPROVED** [review]

**Review by CaiCandong — APPROVED** [review]

**Review by KN4CK3R — APPROVED** [review]

**Comment by wxiaoguang** [comment]
Wait, is it right to get all changed file between current commit and the empty tree commit?

**Comment by KN4CK3R** [comment]
I think the solution is ok but maybe the check should not be in that method but some frames up the callstack.

**Comment by wxiaoguang** [comment]
> The solution could be a problem because this will now scan the whole repo for (dis)allowed files and that may conflict with already existing files which are not modified in the new branch.

Yup, I think so too.

And I recalled this PR:

* #26290

**Comment by lunny** [comment]
> > The solution could be a problem because this will now scan the whole repo for (dis)allowed files and that may conflict with already existing files which are not modified in the new branch.
> 
> Yup, I think so too.
> 
> And I recalled this PR (not fully related but the "empty sha" behavior should be explicitly clarified)
> 
>     * [Support getting changed files when commit ID is `EmptySHA` #26290](https://github.com/go-gitea/gitea/pull/26290)

Maybe we need to check all possbile places which contains oldcommit and newcommit and handle them like #26290 did.

**Comment by silverwind** [comment]
Merge this?

**Comment by lunny** [comment]
This should be replaced by #26290 . @Zettat123 @lng2020

**Comment by lunny** [comment]
Close as outdated.

**Pull request closed without merging**